File: Java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import static java.lang.Integer.*;
import static java.lang.Integer.MIN_VALUE;
import static java.util.function.Function.identity;
import static org.junit.Assert.*;

import org.approvaltests.legacycode.LegacyApprovals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.JunitReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.Test;

import java.util.Arrays;
import java.util.function.Function;
import java.util.stream.Stream;

@UseReporter(JunitReporter.class)
public class GildedRoseTest {

    Stream<String> ODD_ITEM_NAMES = Σ("", "pie", "foo");
    Stream<String> KNOWN_ITEM_NAMES =
         Σ("Aged Brie",
           "Elixir of the Mongoose",
           "Sulfuras, Hand of Ragnaros",
           "Backstage passes to a TAFKAL80ETC concert");
    Stream<String> UNKNOWN_ITEM_NAMES = Σ("+5 Dexterity Vest", "Conjured Mana Cake");

    Stream<Integer> MIN_DAYS =                Σ(MIN_VALUE);
    Stream<Integer> DAYS_NEAR_ZERO =          Σ(-1,0,1);
    Stream<Integer> SINGLE_DIGIT_DAYS =       Σ(4,5,6);
    Stream<Integer> LOW_DOUBLE_DIGIT_DAYS =   Σ(10,11,12);
    Stream<Integer> HIGH_DOUBLE_DIGIT_DAYS =  Σ(99_999,100_000,100_001);
    Stream<Integer> MAX_DAYS =                Σ(MAX_VALUE);

    Stream<Integer> MINIMUM_QUALITY =         Σ(0);
    Stream<Integer> SINGLE_DIGIT_QUALITIES =  Σ(1,2,3,4,5);
    Stream<Integer> DOUBLE_DIGIT_QUALITIES =  Σ(12,13,15);
    Stream<Integer> QUALITIES_NEAR_MAXIMUM =  Σ(47,48,49);
    Stream<Integer> MAXIMUM_QUALITY =         Σ(50);
    Stream<Integer> LEGENDARY_ITEM_QUALITY =  Σ(80);

    Stream<String>  GOOD_MIX_OF_ITEM_NAMES =
            ⁀(ODD_ITEM_NAMES, KNOWN_ITEM_NAMES, UNKNOWN_ITEM_NAMES);
    Stream<Integer> GOOD_MIX_OF_NUMBER_OF_DAYS_BY_WHICH_TO_SELL_ITEM =
            ⁀(MIN_DAYS, DAYS_NEAR_ZERO, SINGLE_DIGIT_DAYS, LOW_DOUBLE_DIGIT_DAYS, HIGH_DOUBLE_DIGIT_DAYS, MAX_DAYS);
    Stream<Integer> GOOD_MIX_OF_ITEM_QUALITY_LEVELS =
            ⁀(MINIMUM_QUALITY, SINGLE_DIGIT_QUALITIES, DOUBLE_DIGIT_QUALITIES, QUALITIES_NEAR_MAXIMUM, MAXIMUM_QUALITY, LEGENDARY_ITEM_QUALITY);

    @Test
    public void foo() {
        String name = "foo";
        int sellIn = 0;
        int quality = 0;
        GildedRose app = updateQuality(name, sellIn, quality);
        assertEquals("fixme", app.items[0].name);
    }

    @Test
    public void regression_test() throws Exception
    {
        String[] names = GOOD_MIX_OF_ITEM_NAMES.toArray(String[]::new);
        Integer[] sellIn = GOOD_MIX_OF_NUMBER_OF_DAYS_BY_WHICH_TO_SELL_ITEM.toArray(Integer[]::new);
        Integer[] quality = GOOD_MIX_OF_ITEM_QUALITY_LEVELS.toArray(Integer[]::new);
        LegacyApprovals.LockDown(this, "updateQuality", names, sellIn, quality);
    }

    public GildedRose updateQuality(String name, Integer sellIn, Integer quality)
    {
        Item[] items = new Item[] { new Item(name, sellIn, quality) };
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        return app;
    }

    // Create a stream of As
    private <A> Stream<A> Σ(A... As){ return Arrays.stream(As); }

    // concatenate streams of As
    private <A> Stream<A> ⁀(Stream<A>... streams) { return Arrays.stream(streams).flatMap( Function.identity() ); }

}

File: go/gildedrose/gildedrose_test.go
package gildedrose_test

import (
	"testing"

	"github.com/emilybache/gildedrose-refactoring-kata/gildedrose"
)

func Test_Foo(t *testing.T) {
	// Arrange
	var items = []*gildedrose.Item{
		{"foo", 0, 0},
	}
	expected := []*gildedrose.Item{
		{"foo", -1, 0},
	}

	// Act
	gildedrose.UpdateQuality(items)

	// Assert
	if items[0].Name != expected[0].Name {
		t.Errorf("Name: Expected %s but got %s ", expected[0].Name, items[0].Name)
	}
	if items[0].SellIn != expected[0].SellIn {
		t.Errorf("SellIn: Expected %d but got %d ", expected[0].SellIn, items[0].SellIn)
	}
	if items[0].Quality != expected[0].Quality {
		t.Errorf("Quality: Expected %d but got %d ", expected[0].Quality, items[0].Quality)
	}
}

func Test_oneDay(t *testing.T) {
	// Arrange
	var items = []*gildedrose.Item{
		{"+5 Dexterity Vest", 10, 20},
		{"Aged Brie", 2, 0},
		{"Elixir of the Mongoose", 5, 7},
		{"Sulfuras, Hand of Ragnaros", 0, 80},
		{"Sulfuras, Hand of Ragnaros", -1, 80},
		{"Backstage passes to a TAFKAL80ETC concert", 15, 20},
		{"Backstage passes to a TAFKAL80ETC concert", 10, 49},
		{"Backstage passes to a TAFKAL80ETC concert", 5, 49},
		{"Conjured Mana Cake", 3, 6},
	}

	var expected = []*gildedrose.Item{
		{"+5 Dexterity Vest", 9, 19},
		{"Aged Brie", 1, 1},
		{"Elixir of the Mongoose", 4, 6},
		{"Sulfuras, Hand of Ragnaros", 0, 80},
		{"Sulfuras, Hand of Ragnaros", -1, 80},
		{"Backstage passes to a TAFKAL80ETC concert", 14, 21},
		{"Backstage passes to a TAFKAL80ETC concert", 9, 50},
		{"Backstage passes to a TAFKAL80ETC concert", 4, 50},
		{"Conjured Mana Cake", 2, 5}, // <-- :O
	}

	// Act
	gildedrose.UpdateQuality(items)

	// Assert
	for i, item := range expected {
		if items[i].Name != item.Name {
			t.Errorf("Name: Expected %s but got %s ", item.Name, items[i].Name)
		}
		if items[i].SellIn != item.SellIn {
			t.Errorf("%s - SellIn: Expected %d but got %d ", item.Name, item.SellIn, items[i].SellIn)
		}
		if items[i].Quality != item.Quality {
			t.Errorf("%s - Quality: Expected %d but got %d ", item.Name, item.Quality, items[i].Quality)
		}
	}

}

func Test_twoDays(t *testing.T) {
	// Arrange
	var items = []*gildedrose.Item{
		{"+5 Dexterity Vest", 10, 20},
		{"Aged Brie", 2, 0},
		{"Elixir of the Mongoose", 5, 7},
		{"Sulfuras, Hand of Ragnaros", 0, 80},
		{"Sulfuras, Hand of Ragnaros", -1, 80},
		{"Backstage passes to a TAFKAL80ETC concert", 15, 20},
		{"Backstage passes to a TAFKAL80ETC concert", 10, 49},
		{"Backstage passes to a TAFKAL80ETC concert", 5, 49},
		{"Conjured Mana Cake", 3, 6},
	}

	var expected = []*gildedrose.Item{
		{"+5 Dexterity Vest", 8, 18},
		{"Aged Brie", 0, 2},
		{"Elixir of the Mongoose", 3, 5},
		{"Sulfuras, Hand of Ragnaros", 0, 80},
		{"Sulfuras, Hand of Ragnaros", -1, 80},
		{"Backstage passes to a TAFKAL80ETC concert", 13, 22},
		{"Backstage passes to a TAFKAL80ETC concert", 8, 50},
		{"Backstage passes to a TAFKAL80ETC concert", 3, 50},
		{"Conjured Mana Cake", 1, 4}, // <-- :O
	}

	// Act
	gildedrose.UpdateQuality(items)
	gildedrose.UpdateQuality(items)

	// Assert
	for i, item := range expected {
		if items[i].Name != item.Name {
			t.Errorf("Name: Expected %s but got %s ", item.Name, items[i].Name)
		}
		if items[i].SellIn != item.SellIn {
			t.Errorf("%s - SellIn: Expected %d but got %d ", item.Name, item.SellIn, items[i].SellIn)
		}
		if items[i].Quality != item.Quality {
			t.Errorf("%s - Quality: Expected %d but got %d ", item.Name, item.Quality, items[i].Quality)
		}
	}
}
func Test_fiveDays(t *testing.T) {
	// Arrange
	var items = []*gildedrose.Item{
		{"+5 Dexterity Vest", 10, 20},
		{"Aged Brie", 2, 0},
		{"Elixir of the Mongoose", 5, 7},
		{"Sulfuras, Hand of Ragnaros", 0, 80},
		{"Sulfuras, Hand of Ragnaros", -1, 80},
		{"Backstage passes to a TAFKAL80ETC concert", 15, 20},
		{"Backstage passes to a TAFKAL80ETC concert", 10, 49},
		{"Backstage passes to a TAFKAL80ETC concert", 5, 49},
		{"Conjured Mana Cake", 3, 6},
	}

	var expected = []*gildedrose.Item{
		{"+5 Dexterity Vest", 0, 10},
		{"Aged Brie", -8, 18},
		{"Elixir of the Mongoose", -5, 0},
		{"Sulfuras, Hand of Ragnaros", 0, 80},
		{"Sulfuras, Hand of Ragnaros", -1, 80},
		{"Backstage passes to a TAFKAL80ETC concert", 5, 35},
		{"Backstage passes to a TAFKAL80ETC concert", 0, 50},
		{"Backstage passes to a TAFKAL80ETC concert", -5, 0},
		{"Conjured Mana Cake", -7, 0}, // <-- :O
	}

	// Act
	for day := 1; day <= 10; day++ {
		gildedrose.UpdateQuality(items)
	}

	// Assert
	for i, item := range expected {
		if items[i].Name != item.Name {
			t.Errorf("Name: Expected %s but got %s ", item.Name, items[i].Name)
		}
		if items[i].SellIn != item.SellIn {
			t.Errorf("%s - SellIn: Expected %d but got %d ", item.Name, item.SellIn, items[i].SellIn)
		}
		if items[i].Quality != item.Quality {
			t.Errorf("%s - Quality: Expected %d but got %d ", item.Name, item.Quality, items[i].Quality)
		}
	}
}
